platform/darwin/src/MGLFeature.mm

```diff
@@ -108,13 +206,47 @@ @implementation MGLShapeCollectionFeature
 @synthesize identifier;
 @synthesize attributes;
 
+@dynamic shapes;
+
++ (instancetype)shapeCollectionWithShapes:(NSArray *)shapes {
+    return [super shapeCollectionWithShapes:shapes];
+}
+
 - (id)attributeForKey:(NSString *)key {
     return self.attributes[key];
 }
 
-@end
+- (NSDictionary *)featureDictionary {
+    return @{@"type": @"Feature",
+             @"properties": (self.attributes) ? self.attributes : @{},
+             @"geometry": @{@"type": @"GeometryCollection",
+                            @"geometries": [self geometryCollection]}};
+}
+
+- (NSArray *)geometryCollection {
+    NSMutableArray *geometries = [NSMutableArray array];
+    
+    for (MGLShape<MGLFeature> *shape in self.shapes) {
+        if ([shape isKindOfClass:[MGLShapeCollectionFeature class]]) {
+            [geometries addObject:@{@"type": @"GeometryCollection",
+                                    @"geometries": [self geometryCollection]}];
+        } else {
+            NSDictionary *geometry = shape.featureDictionary[@"geometry"];
+            [geometries addObject:@{@"type": geometry[@"type"],
+                                    @"coordinates": geometry[@"coordinates"] }];
+        }
+    }
+    
+    return [geometries copy];
```

Comment:
This is intentional. The method says that it returns a non-mutable array even though it uses a mutable array internally to create the collection

> (NSArray *)geometryCollection